src/llm/wmt_data_utils.py

```diff
@@ -103,59 +97,85 @@ def decode_wmt_tokens(
 
 
-def _collate_fn(batch, source, target, tokenizers):
+def _collate_fn(batch, source, target, tokenizer, pad_id=0):
     """
     Collate function for the dataloader. This will collate the batches into a single tensor.
 
     Args:
         batch (list): The batch of data.
         source (str): The source language.
         target (str): The target language.
-        tokenizers (dict): The tokenizers for the specified dataset name.
+        tokenizer (Tokenizer): The tokenizer.
+        pad_id (int): The padding id.
 
     Returns:
         tuple: The collated batches.
     """
     zipped_list = [
         (
             sample["translation"][source],
-            sample["translation"][target],
+            f"[BOS] {sample['translation'][target]} [EOS]",
         )
         for sample in batch
     ]
     source_batch, target_batch = list(zip(*zipped_list))
 
-    # enable padding for the tokenizers
-    # we do the longest sequence length for padding
-    tokenizers[source].enable_padding(pad_id=1, pad_token="<pad>")
-    tokenizers[target].enable_padding(pad_id=1, pad_token="<pad>")
-
     # tokenize the batches
     tokenized_source_batch = [
-        token.ids for token in tokenizers[source].encode_batch(source_batch)
+        token.ids for token in tokenizer.encode_batch(source_batch)
     ]
     tokenized_target_batch = [
-        token.ids for token in tokenizers[target].encode_batch(target_batch)
-    ]
-    # Add start token (<s>) to beginning of each target sequence to shift tokens right by 1.
-    tokenized_target_batch = [
-        [tokenizers[target].token_to_id("<s>")] + token
-        for token in tokenized_target_batch
+        token.ids for token in tokenizer.encode_batch(target_batch)
     ]
 
     # torch tensors
     tokenized_source_batch = torch.tensor(tokenized_source_batch)
     tokenized_target_batch = torch.tensor(tokenized_target_batch)
 
-    return tokenized_source_batch, tokenized_target_batch
+    # Encoder attention mask
+    encoder_attention_mask = (tokenized_source_batch != pad_id).bool()
+    # [batch, enc_seq_len] -> [batch, enc_seq_len, enc_seq_len]
+    B, enc_seq_len = encoder_attention_mask.size()
+    encoder_self_attention_mask = encoder_attention_mask.unsqueeze(1).expand(
+        B, enc_seq_len, enc_seq_len
+    )
+
+    # Decoder attention mask
+    decoder_attention_mask = (tokenized_target_batch != pad_id).bool()
+    # [batch, dec_seq_len] -> [batch, dec_seq_len, dec_seq_len]
+    _, dec_seq_len = decoder_attention_mask.size()
+    decoder_self_attention_mask = decoder_attention_mask.unsqueeze(1).expand(
+        B, dec_seq_len, dec_seq_len
+    )
+
+    # Decoder cross attention mask
+    decoder_attention_mask = decoder_attention_mask.unsqueeze(2)  # [batch, dec_len, 1]
+    encoder_attention_mask = encoder_attention_mask.unsqueeze(1)  # [batch, 1, enc_len]
+    decoder_cross_attention_mask = (
+        decoder_attention_mask & encoder_attention_mask
+    )
```

Comment:
I think we can just exploit tokenizer's `.mask` api here? WDYT?

Reply:
can you share documentation or working code?

Reply:
working on this.

Reply:
https://colab.research.google.com/drive/17m1ZwTZrmql1LHDtRA5BRWSXgPccSw5U?usp=sharing